### PR TITLE
Adding DQ_SerialManipulatorDH 

### PR DESCRIPTION
**Pull request by mmmarinho**
This pull request adds the `DQ_SerialManipulatorDH` class.
This update includes the last features intended for version 20.04.

### CODE REVIEW

**Comment by bvadorno** [comment]
Hi, @mmmarinho,

Everything looks fine, although I have some comments and questions concerning the new `DQ_SerialManipulatorDH.m`

1. If the class only accepts the standard convention, why do we need to have a second parameter in the constructor to determine the DH convention in the class? If the goal is to ensure temporary compatibility with the current implementation, I suggest doing the following:
```
function obj = DQ_SerialManipulatorDQ(A,~)
    % do whatever it has to do
end
```
This way, if one calls the constructor with two parameters, the second one is just ignored. In the case of C++, it suffices to design a constructor with just one parameter and overload it with a constructor that uses two parameters. The overloading constructor would then just call the main one while ignoring the second parameter.

2. Since there are no dummy joints in `DQ_SerialManipulator`, I suggest removing it from the comments throughout the `DQ_SerialManipulatorDH` class.

3. When considering the previous implementation in the `DQ_SerialManipulator` class, since the only difference between the optimized version of the method `dh2dq`, _in terms of optimizations_, consists in declaring variables such as `sine_of_half_theta`, etc., I suggest removing the comments related to the non-optimized implementation altogether and adding the bibliographic reference related to the implementation.

    3.1 Could you also break the lines at 80 characters to comply with our current guide style?

4. In the method `raw_pose_jacobian`, how is the current optimization better than the previous implementation in the `DQ_SerialManipulator` class? In the previous one, we needed essentially two dual quaternion multiplications, whereas in the current optimized version, we need four. Could you please clarify this point?

Cheers,
Bruno

**Comment by mmmarinho** [comment]
Hello, @bvadorno,

Thank you for the review.
 
>     1. If the class only accepts the standard convention, why do we need to have a second parameter in the constructor to determine the DH convention in the class? If the goal is to ensure temporary compatibility with the current implementation, I suggest doing the following:

I wanted to check with you today if we are going to make the modified DH parameters a different class or going to use the `DQ_SerialManipulatorDH` class. In the former case, we can temporarily follow your proposed strategy (and fade it out in 20.10); in the latter, I can add the implementation for the modified parameters in 20.04.
> 
>     1. Since there are no dummy joints in `DQ_SerialManipulator`, I suggest removing it from the comments throughout the `DQ_SerialManipulatorDH` class.
Good catch, thanks.

>     2. When considering the previous implementation in the `DQ_SerialManipulator` class, since the only difference between the optimized version of the method `dh2dq`, _in terms of optimizations_, consists in declaring variables such as `sine_of_half_theta`, etc., I suggest removing the comments related to the non-optimized implementation altogether and adding the bibliographic reference related to the implementation.
Sure thing.

>        3.1 Could you also break the lines at 80 characters to comply with our current guide style?
No problem.

>     3. In the method `raw_pose_jacobian`, how is the current optimization better than the previous implementation in the `DQ_SerialManipulator` class? In the previous one, we needed essentially two dual quaternion multiplications, whereas in the current optimized version, we need four. Could you please clarify this point?

Trying to remember the profiling results off the top of my head, the implementation in `DQ_SerialManipulatorDH` is about 20% slower than the one in `DQ_SerialManipulator` in MATLAB. Just to add a bit to this topic, the profiling results show that, as expected, the problem isn't the extra multiplications themselves, but that new DQs have to be instanced and memory operations in MATLAB are a bit slow. This is only to say that I expect the overall results to be less pronounced in CPP/Python.

I wasn't aiming to further optimize the method. I think you optimized it thoroughly in the earlier implementation. I was aiming to make it a bit easier to understand, although the meaning of _easier_ depends on point-of-view (`get_z()` tended to be a difficult point for some students), so I guess this is a good opportunity to discuss this trade-off. 

When I modified the `raw_pose_jacobian`, the main point was to make `dh2dq_dot()` explicit. The idea behind doing this is that we can follow a similar procedure in other concrete implementations of the (eventually, purely abstract) `DQ_SerialManipulator`, by having a `parameter_to_dq()` and a `parameter_to_dq_dot()`. What do you think?

**Comment by mmmarinho** [comment]
Hello, @bvadorno,

Just to update this thread with the conversation we had yesterday.

> I wanted to check with you today if we are going to make the modified DH parameters a different class or going to use the `DQ_SerialManipulatorDH` class. In the former case, we can temporarily follow your proposed strategy (and fade it out in 20.10); in the latter, I can add the implementation for the modified parameters in 20.04.

I forgot to mention this yesterday. I'm personally pending towards making a different class for each representation (because it reduces the number of if-else in a given code and makes it easier to maintain the code). Please let me know what you think. 

> (...) When I modified the `raw_pose_jacobian`, the main point was to make `dh2dq_dot()` explicit. The idea behind doing this is that we can follow a similar procedure in other concrete implementations of the (eventually, purely abstract) `DQ_SerialManipulator`, by having a `parameter_to_dq()` and a `parameter_to_dq_dot()`. What do you think?

Thank you for the explanation yesterday. Given all the theoretical development put into Algorithm 2.1 ([1], Page 41, based on (2.25) to (2.30)) and considering all development that was made based on it, I have to say that I'm fully convinced to keep using it. I'll rewrite `DQ_SerialManipulatorDH` to use it. 

We can consider, in the future, having the main algorithm that calculates the `raw_pose_jacobian()` in `DQ_Kinematics`. Each subclass would implement, then, the `parameter_to_dq()` method and the `get_z()` (probably renamed to something else, for which you're more qualified than me to choose). =) 
This understanding coincides with "The general procedure to calculate the dual quaternion Jacobian convention is summarized in Algorithm 2.1. The calculations in lines 4 and 6 (of Algorithm 2.1) depend on the set of parameters used (...)".

For 20.04, we might consider just renaming `get_z()`.

[[1]](https://adorno.eng.ufmg.br/publications/phd_thesis_final_version.pdf) ADORNO, B. V., Two-arm manipulation: from manipulators to enhanced human-robot collaboration [Contribution à la manipulation à deux bras : des manipulateurs à la collaboration homme-robot], Université Montpellier 2, Montpellier, France, 2011

File: robot_modeling/DQ_SerialManipulatorDH.m [hunk 1]
```diff
@@ -283,31 +228,17 @@
...
-                x_backward = conj(x_i_to_i_plus_one)*x_backward;
-                x_ith_dot  = x_forward*obj.dh2dq_dot(q(i),i)*x_backward;
-                x_forward  = x_forward*x_i_to_i_plus_one;
-                J(:,i) = vec8(x_ith_dot);
+            for i = 0:to_ith_link-1
+                w = obj.get_w(i+1);
+                z = 0.5*x*w*x';
+                x = x*obj.dh2dq(q(i+1),i+1);
+                j = z * x_effector;
+                J(:,i+1) = vec8(j);
```
Comment (mmmarinho) [inline robot_modeling/DQ_SerialManipulatorDH.m:241]: Test

**Comment by bvadorno** [comment]
> Hello, @bvadorno,
> 
> Just to update this thread with the conversation we had yesterday.
> 
> > I wanted to check with you today if we are going to make the modified DH parameters a different class or going to use the `DQ_SerialManipulatorDH` class. In the former case, we can temporarily follow your proposed strategy (and fade it out in 20.10); in the latter, I can add the implementation for the modified parameters in 20.04.
> 
> I forgot to mention this yesterday. I'm personally pending towards making a different class for each representation (because it reduces the number of if-else in a given code and makes it easier to maintain the code). Please let me know what you think.
> 

I agree with you. A different class for each representation seems the best approach.

> > (...) When I modified the `raw_pose_jacobian`, the main point was to make `dh2dq_dot()` explicit. The idea behind doing this is that we can follow a similar procedure in other concrete implementations of the (eventually, purely abstract) `DQ_SerialManipulator`, by having a `parameter_to_dq()` and a `parameter_to_dq_dot()`. What do you think?
> 
> Thank you for the explanation yesterday. Given all the theoretical development put into Algorithm 2.1 ([1], Page 41, based on (2.25) to (2.30)) and considering all development that was made based on it, I have to say that I'm fully convinced to keep using it. I'll rewrite `DQ_SerialManipulatorDH` to use it.
> 

Great!

> We can consider, in the future, having the main algorithm that calculates the `raw_pose_jacobian()` in `DQ_Kinematics`. Each subclass would implement, then, the `parameter_to_dq()` method and the `get_z()` (probably renamed to something else, for which you're more qualified than me to choose). =)
> This understanding coincides with "The general procedure to calculate the dual quaternion Jacobian convention is summarized in Algorithm 2.1. The calculations in lines 4 and 6 (of Algorithm 2.1) depend on the set of parameters used (...)".
> 

Great idea, @mmmarinho! However, since `DQ_SerialManipulator` will become an abstract class in the near future, I think it would be the right place to put it. The main reason is that Algorithm 2.1 applies only to serial manipulators, whereas `DQ_Kinematics` is a more general abstract class. What do you think?


> [[1]](https://adorno.eng.ufmg.br/publications/phd_thesis_final_version.pdf) ADORNO, B. V., Two-arm manipulation: from manipulators to enhanced human-robot collaboration [Contribution à la manipulation à deux bras : des manipulateurs à la collaboration homme-robot], Université Montpellier 2, Montpellier, France, 2011

Cheers,
Bruno

**Comment by mmmarinho** [comment]
> Great idea, @mmmarinho! However, since `DQ_SerialManipulator` will become an abstract class in the near future, I think it would be the right place to put it. The main reason is that Algorithm 2.1 applies only to serial manipulators, whereas `DQ_Kinematics` is a more general abstract class. What do you think?

Hello, @bvadorno. You're right, `DQ_SerialManipulator` would be more suitable. I wrote `DQ_Kinematics` by mistake. :)

Kind regards,
Murilo

**Pull request merged**